api/catalog/api/serializers/media_serializers.py

```diff
@@ -73,6 +86,61 @@ class TagSerializer(serializers.Serializer):
     )
 
 
+def get_search_request_source_serializer(media_type):
+    class MediaSearchRequestSourceSerializer(serializers.Serializer):
+        """
+        This serializer parses and validates the source/not_source fields from the query
+        parameters.
+        """
+
+        field_names = [
+            "source",
+            "excluded_source",
+        ]
+        """
+        Keep the fields names in sync with the actual fields below as this list is
+        used to generate Swagger documentation.
+        """
+
+        _field_attrs = {
+            "help_text": (
+                "A comma separated list of data sources to search. "
+                f"Valid inputs: {format_enums(get_sources(media_type).keys())}"
+            ),
+            "required": False,
+        }
+
+        source = serializers.CharField(label="provider", **_field_attrs)
+        excluded_source = serializers.CharField(
+            label="excluded_provider", **_field_attrs
+        )
+
+        @staticmethod
+        def validate_source_field(input_sources):
+            allowed_sources = list(get_sources(media_type).keys())
+            input_sources = input_sources.split(",")
+            input_sources = [x for x in input_sources if x in allowed_sources]
+            input_sources = ",".join(input_sources)
+            return input_sources.lower()
+
+        def validate_source(self, input_sources):
+            return self.validate_source_field(input_sources)
+
+        def validate_excluded_source(self, input_sources):
+            return self.validate_source(input_sources)
+
+        def validate(self, data):
+            data = super().validate(data)
+            if "source" in self.initial_data and "excluded_source" in self.initial_data:
+                raise serializers.ValidationError(
+                    "Cannot set both 'source' and 'excluded_source'. "
+                    "Use exactly one of these."
+                )
+            return data
+
+    return MediaSearchRequestSourceSerializer
```

Comment:
It just occurred to me that the class name for any serializer created with this factory function will be the same, no?

Should we use the `type` constructor to create a dynamically defined class with an appropriate name? I'm recalling having to do that in the past for decorated class objects that needed their name changed to reflected the decorated status.

I'm not sure if this would effect stack traces or not.

Reply:
While using `type()` will allow us to use dynamic class names, it'll make the code far less readable and much more difficult to understand. I strongly feel that this is not a valuable tradeoff.